packages/@atjson/offset-annotations/src/utils/social-urls.ts

```diff
@@ -350,9 +366,7 @@ function normalizeTelegramUrl(url: IUrl) {
   };
 }
 
-export function identify(
-  url: IUrl
-): {
+export function identify(url: IUrl): {
```

Comment:
So you're seeing a type error here since the constructor for an arbitrary IframeEmbed<T> may now be different, but this is in fact the case since we are returning classes that have a different constructor function now. In the long run, it will probably be easier to provide a more meaningful type here:
```
type EmbedInfo<T> = {
  attributes: AttributesOf<T>;
  Class: new (attr: {
    start: number;
    end: number;
    attributes: AttributesOf<T>;
  }) => T;
};
```
Note that we're typing `Class` specifically as the constructor.

You can create a type for exactly what you are returning here:
```
type SocialEmbedInfo = 
  | EmbedInfo<FacebookEmbed>
  | EmbedInfo<GiphyEmbed>
  | EmbedInfo<RedditEmbed>
  ...
  | null;

export function identify(url: IUrl): SocialEmbedInfo
```

Why do all of this? This will allow you to include RedditEmbed-specific types in your normalize function above:
```
function normalizeRedditURL(url: IUrl): EmbedInfo<RedditEmbed> {
  ...
}
```
And later, when you call this identify function, you will be able to create any type of embed annotation that is typed properly:
```
// somewhere else, like when you are trying to identify a url
const embedInfo = identify(someUrl);
if (embedInfo) {
  const {attributes, Class} = embedInfo;
  const newAnnotation = new Class({
    start,
    end,
    attributes
  }); // typescript will know that newAnnotation is a FacebookEmbed | GiphyEmbed | RedditEmbed | ...
}
```

Reply:
Thanks for the explanation @bachbui, i have made the code changes, please have a look on it.